src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -2457,6 +2472,11 @@ private void OnCompleted(Action continuation, bool flowExecutionContext)
 
                 try
                 {
+                    if (this.Kind == LockKind.Read)
+                    {
+                        this.continuationTaskScheduler = this.OwningLock.GetTaskSchedulerForLockRequest();
```

Comment:
Why store this in a field here, and use it later rather than just make this method call when we need the value?

Reply:
The logic behind it is that we want to use a special scheduler, only when it is inside a UI thread blocking call.  That condition can only be accessed during the initialization phase, when it is still in the ExecutionContext of calling ReadAsync().  When we issue the lock, we only return to the context when the task is resumed, which will happen later than the scheduler is used.